### PR TITLE
Adds support to access mayactl with cluster ip and refactors code to increase modularity 

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

**What this PR does / why we need it**: 
-> This pr adds support for accessing mayactl outside the pod inside the master with cluster IP (Currently not working in google cloud).
-> Refactors the code to reduce complexity and meet go standards.
-> Adds get and post method in util.go to avoid repetitive steps.
-> Adds unit tests to improve code coverage.

**Special notes for your reviewer**:
-> Removed redundant checking of mapiaddr in each function and added the check in command.go .
-> Merge this PR after merging this pr https://github.com/openebs/maya/pull/370 .
screenshot from 2018-05-31 22-54-25

### CODE REVIEW

File: cmd/mayactl/app/command/commands.go [hunk 1]
```diff
@@ -16,17 +16,32 @@ package command
...
 )
 
-// NewCommand creates the `maya` command and its nested children.
+// NewMayaCommand creates the `maya` command and its nested children.
 func NewMayaCommand() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "mayactl",
-		Short: "Maya means 'Magic' a tool for storage orchestration",
-		Long:  `Maya means 'Magic' a tool for storage orchestration`,
+		Short: "Mayactl means 'Magic'a tool for storage orchestration",
```
Comment (prateekpandey14) [inline cmd/mayactl/app/command/commands.go]: Maya means magic :smile:

File: cmd/mayactl/app/command/volume.go [hunk 2]
```diff
@@ -88,3 +93,25 @@ func NewCmdVolume() *cobra.Command {
...
+// no error and returns an error if it is missing.
+func (c *CmdVolumeOptions) Validate(cmd *cobra.Command, snapshotnameverify, sourcenameverify, volnameverify bool) error {
+	if snapshotnameverify {
+		if len(c.snapshotName) == 0 {
+			return errors.New("--snapshotname is missing. Please provide a snapshotname")
+		}
+	}
+	if sourcenameverify {
+		if len(c.sourceVolumeName) == 0 {
+			return errors.New("--sourcevolumename is missing. Please specify a sourcename")
```
Comment (prateekpandey14) [inline cmd/mayactl/app/command/volume.go]: `sourcevolume`  ?

File: pkg/client/mapiserver/snapshot_test.go [hunk 3]
```diff
@@ -357,6 +328,7 @@ func TestListSnapshot(t *testing.T) {
 			defer os.Unsetenv(tt.addr)
 			defer server.Close()
 			got := ListSnapshot(tt.volumeName, tt.namespace)
+			// t.Errorf("================%v===================\n%+v\n%+v\n******************", name, got, tt.err)
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot_test.go]: remove this

File: pkg/client/mapiserver/volume_delete.go [hunk 4]
```diff
@@ -14,31 +10,6 @@ const (
...
-		return err
-	}
-
-	defer resp.Body.Close()
-	code := resp.StatusCode
-	if code != http.StatusOK {
-		return fmt.Errorf("Status error: %v ", http.StatusText(code))
-	}
-	return nil
+	_, err := getRequest(GetURL()+"/latest/volumes/delete/"+vname, namespace, false)
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/volume_delete.go]: can we define go constant for endpoints ? i.e. deleteEndpoint, createEndpoint ?

File: cmd/mayactl/app/command/commands.go (hunk 1 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/commands.go]: ok

File: cmd/mayactl/app/command/volume.go (hunk 2 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/volume.go]: yup fogot to change that.

File: pkg/client/mapiserver/snapshot_test.go (hunk 3 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/snapshot_test.go]: ok

File: pkg/client/mapiserver/volume_delete.go (hunk 4 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/volume_delete.go]: yup we can

Comment (ashishranjan738) [reply pkg/client/mapiserver/volume_delete.go]: but i think we should not do it in this pr

File: cmd/mayactl/app/command/commands.go [hunk 5]
```diff
@@ -16,17 +16,32 @@ package command
...
-// NewCommand creates the `maya` command and its nested children.
+// NewMayaCommand creates the `maya` command and its nested children.
 func NewMayaCommand() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "mayactl",
-		Short: "Maya means 'Magic' a tool for storage orchestration",
+		Short: "Maya means 'Magic'a tool for storage orchestration",
 		Long:  `Maya means 'Magic' a tool for storage orchestration`,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			if mapiserver.MAPIAddr == "" {
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/commands.go]: can you make it len(mapiserver.MAPIAddr) == 0  , doing this way makes code clear.

File: cmd/mayactl/app/command/snapshot/create.go [hunk 6]
```diff
@@ -35,25 +35,6 @@ $ mayactl snapshot create --volname <vol> --snapname <snap>
 `
 )
 
-/*func init() {
-	host := os.Getenv("MAPI_ADDR")
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/snapshot/create.go]: Is this code will not be used later ? Please verify with @prateekpandey14

File: cmd/mayactl/app/command/stats_helper.go [hunk 7]
```diff
@@ -4,13 +4,15 @@ import (
...
-	"os"
 	"time"
 
+	"github.com/openebs/maya/pkg/client/mapiserver"
+
 	"github.com/openebs/maya/pkg/util"
 	"github.com/openebs/maya/types/v1"
 )
 
+// Client interface contans GetVolAnnotation
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/stats_helper.go]: Can comment be changed to "Client interface defines the GetVolAnnotation method which can be used to get the details about the volumes from maya-apiserver." ?

File: cmd/mayactl/app/command/stats_helper.go [hunk 8]
```diff
@@ -32,16 +34,10 @@ const (
...
 func GetVolDetails(volName string, namespace string, obj interface{}) error {
-	addr := os.Getenv("MAPI_ADDR")
-	if addr == "" {
-		err := util.MAPIADDRNotSet
-		fmt.Printf("error getting env variable: %v", err)
-		return err
-	}
 
-	url := addr + "/latest/volumes/info/" + volName
+	url := mapiserver.GetURL() + "/latest/volumes/info/" + volName
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/stats_helper.go]: Please give some comments why this utility function is needed and move "/latest/volumes/info" to constant var "InfoAPIEndpoint".

File: cmd/mayactl/app/command/stats_helper_test.go [hunk 9]
```diff
@@ -63,16 +63,6 @@ func TestGetVolDetails(t *testing.T) {
 			err:  nil,
 			addr: "MAPI_ADDR",
 		},
-		"MAPIADDRNotSet": {
-			volumeName: "vol",
```
Comment (utkarshmani1997) [inline cmd/mayactl/app/command/stats_helper_test.go]: What will be default scenario, how are you handling that ?
is there any test case for such scenario ?

File: pkg/client/k8s/k8s.go [hunk 10]
```diff
@@ -570,7 +571,8 @@ func (k *K8sClient) DeleteAppsV1B1Deployment(name string) error {
 
 func getK8sConfig() (config *rest.Config, err error) {
 	k8sMaster := api_oe_old.K8sMasterENV()
-	kubeConfig := api_oe_old.KubeConfigENV()
+	// kubeConfig := api_oe_old.KubeConfigENV()
+	kubeConfig := os.Getenv("KUBECONFIG")
```
Comment (utkarshmani1997) [inline pkg/client/k8s/k8s.go]: can you please mention why you commented the above code and why have you used os.Getenv()? I believe there are certain scenarios for which the commented code stands valid.

File: pkg/client/mapiserver/snapshot.go [hunk 11]
```diff
@@ -49,135 +46,53 @@ type SnapshotInfo struct {
...
-	_, err = ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return err
-	}
-	code := resp.StatusCode
-
-	if code != http.StatusOK {
-		return fmt.Errorf("Server status error: %v", http.StatusText(code))
-	}
-	return nil
```
Comment (utkarshmani1997) [inline pkg/client/mapiserver/snapshot.go]: can you please mention why this code has been removed in PR template, i find it difficult to understand.

File: pkg/client/mapiserver/snapshot.go [hunk 12]
```diff
@@ -49,135 +46,53 @@ type SnapshotInfo struct {
...
+			Name: snapName,
+		},
+		Spec: v1.VolumeSnapshotSpec{
+			VolumeName: volName,
+		},
+	}
+
+	// Marshal serializes the values
+	jsonValue, _ := json.Marshal(snap)
+	_, err := postRequest(GetURL()+"/latest/snapshots/revert/", jsonValue, namespace, false)
```
Comment (utkarshmani1997) [inline pkg/client/mapiserver/snapshot.go]: is it good to ignore one of the value, what does it returns ?

File: pkg/client/mapiserver/snapshot.go [hunk 13]
```diff
@@ -49,135 +46,53 @@ type SnapshotInfo struct {
...
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return err
-	}
-
-	code := resp.StatusCode
-	if code != http.StatusOK {
-		return fmt.Errorf("Server status error: %v", http.StatusText(code))
-	}
```
Comment (utkarshmani1997) [inline pkg/client/mapiserver/snapshot.go]: i thinks you have written some utility function for this repetitive task, that's good.

File: cmd/mayactl/app/command/commands.go (hunk 5 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/commands.go]: sure

File: cmd/mayactl/app/command/snapshot/create.go (hunk 6 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/snapshot/create.go]: i think i had discussed that

File: cmd/mayactl/app/command/stats_helper.go (hunk 7 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/stats_helper.go]: ok sure

File: cmd/mayactl/app/command/stats_helper.go (hunk 8 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/stats_helper.go]: ok

File: cmd/mayactl/app/command/stats_helper_test.go (hunk 9 above)
Comment (ashishranjan738) [reply cmd/mayactl/app/command/stats_helper_test.go]: Its getting handled in command.go file I think handling it everywhere is just repetition

File: pkg/client/k8s/k8s.go (hunk 10 above)
Comment (ashishranjan738) [reply pkg/client/k8s/k8s.go]: it is because on running mayactl outside the pod we will be needing  kubeconfig to make request from kube api-server

File: pkg/client/mapiserver/snapshot.go (hunk 12 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/snapshot.go]: because for the current post request we will not be getting any response from server .

File: pkg/client/mapiserver/snapshot.go (hunk 13 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/snapshot.go]: yup get and post function has been added to modularize the code in util.go

**Comment by ashishranjan738** [comment]
made the requested changes .

File: pkg/client/mapiserver/snapshot.go (hunk 12 above)
Comment (prateekpandey14) [inline pkg/client/mapiserver/snapshot.go]: Handle the marshal err here please....

File: pkg/client/mapiserver/utils.go [hunk 14]
```diff
@@ -62,3 +76,89 @@ func ChangeDateFormatToUnixDate(snapshotDisks []SnapshotInfo) error {
...
+
+	req.Header.Add("Content-Type", "application/json")
+
+	if len(namespace) > 0 {
+		req.Header.Set("namespace", namespace)
+	}
+
+	c := &http.Client{
+		Timeout: volumeCreateTimeout,
+	}
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/utils.go]: timeout constant are not defined ..?

File: pkg/client/mapiserver/utils.go [hunk 15]
```diff
@@ -62,3 +76,89 @@ func ChangeDateFormatToUnixDate(snapshotDisks []SnapshotInfo) error {
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(namespace) > 0 {
+		req.Header.Set("namespace", namespace)
+	}
+
+	c := &http.Client{
+		Timeout: timeoutVolumeDelete,
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/utils.go]: Can we have a single timeout constant .... `DefaultVolumeTimeout` in all volume related operations. ?

File: pkg/client/mapiserver/snapshot.go (hunk 12 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/snapshot.go]: it is handled in the function that is calling it

File: pkg/client/mapiserver/utils.go (hunk 14 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/utils.go]: it is defined in volume_create.go

File: pkg/client/mapiserver/utils.go (hunk 15 above)
Comment (ashishranjan738) [reply pkg/client/mapiserver/utils.go]: yes we can have it

File: pkg/client/k8s/k8s.go (hunk 10 above)
Comment (prateekpandey14) [reply pkg/client/k8s/k8s.go]: Handle this part in case of using external kubeconfig from outside the maya-apiserver pod.

File: pkg/client/mapiserver/volume_create.go [hunk 16]
```diff
@@ -1,66 +1,33 @@
...
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: vname,
+		},
+		Capacity: size,
 	}
 
-	req.Header.Add("Content-Type", "application/yaml")
+	// Marshal serializes the value of vs structure
+	jsonValue, _ := json.Marshal(vs)
```
Comment (prateekpandey14) [inline pkg/client/mapiserver/volume_create.go]: handle this marshal error here

File: cmd/mayactl/app/command/commands.go [hunk 17]
```diff
@@ -16,17 +16,32 @@ package command
...
 	"github.com/spf13/cobra"
 )
 
-// NewCommand creates the `maya` command and its nested children.
+// NewMayaCommand creates the `maya` command and its nested children.
 func NewMayaCommand() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "mayactl",
-		Short: "Maya means 'Magic' a tool for storage orchestration",
+		Short: "Maya means 'Magic'a tool for storage orchestration",
```
Comment (kmova) [inline cmd/mayactl/app/command/commands.go]: n_ fix the punctuate

File: cmd/mayactl/app/command/commands.go [hunk 18]
```diff
@@ -38,6 +53,9 @@ func NewMayaCommand() *cobra.Command {
 	// add the glog flags
 	cmd.PersistentFlags().AddGoFlagSet(flag.CommandLine)
 
+	// add the api addr flag
+	cmd.PersistentFlags().StringVarP(&mapiserver.MAPIAddr, "mapiserver", "m", "", "Address of Maya-Apiserver")
```
Comment (kmova) [inline cmd/mayactl/app/command/commands.go]: Address of Maya-Apiserver -> "Maya API Service IP address.  You can obtain the IP address using `kubectl get svc -n < namespace where openebs is installed >`

Comment (kmova) [reply cmd/mayactl/app/command/commands.go]: Also include a flag for port. With default value specified as 5656.

Comment (ashishranjan738) [reply cmd/mayactl/app/command/commands.go]: ok

File: cmd/mayactl/app/command/commands.go [hunk 19]
```diff
@@ -54,8 +54,8 @@ func NewMayaCommand() *cobra.Command {
 	cmd.PersistentFlags().AddGoFlagSet(flag.CommandLine)
 
 	// add the api addr flag
-	cmd.PersistentFlags().StringVarP(&mapiserver.MAPIAddr, "mapiserver", "m", "", "Address of Maya-Apiserver")
-
+	cmd.PersistentFlags().StringVarP(&mapiserver.MAPIAddr, "mapiserver", "m", "", "Maya API Service IP address. You can obtain the IP address using kubectl get svc -n < namespace where openebs is installed >")
+	cmd.PersistentFlags().StringVarP(&mapiserver.MAPIAddrPort, "mapiserverport", "p", "", "Maya API Service Port.")
```
Comment (prateekpandey14) [inline cmd/mayactl/app/command/commands.go]: Set default port value here, so you don't have to handle in below code.

File: pkg/client/mapiserver/snapshot.go [hunk 20]
```diff
@@ -231,16 +159,8 @@ func getInfo(body []byte) (map[string]client.DiskInfo, error) {
 	return s, err
 }
 
+// displayVolumeSnapshot displays the snapshot information in the snapshot template
 func displayVolumeSnapshot(snapshotList []SnapshotInfo) error {
-	const (
```
Comment (kmova) [inline pkg/client/mapiserver/snapshot.go]: Why is this deleted? Is the template note used?
